File: 377_Combination_Sum_IV/combinationSum4.py
class Solution(object):
    def combinationSum4(self, nums, target):
        self.res = 0
        nums.sort()
        def dfs(nums, remain):
            if not remain:
                self.res += 1
            for num in nums:
                if num > remain: return 
                dfs(nums, remain - num)

        dfs(nums, target)
        return self.res

sol = Solution()
nums = [4, 2, 1]
target = 32
print sol.combinationSum4(nums, target)

File: 377_Combination_Sum_IV/combinationSum4.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
    	int combinationSum4(vector<int>& nums, int target) {
            vector<int> dp(target + 1);
            sort(nums.begin(), nums.end());
            dp[0] = 1;
            for(int i = 1; i <= target; i++) {
                for(int j = 0; j < nums.size(); j++) {
                    if(i >= nums[j])
                        dp[i] += dp[i - nums[j]];
                    else break;
                }
            }
            return dp[target];
    	}
};

int main() {
    Solution sol;
    int arr[] = {4, 2, 1};
    int target = 32;
    vector<int> nums(arr, arr + 3);
    cout << sol.combinationSum4(nums, target) << endl;
}
